File: 例子/Java的例子/pure/SignSEStamp.java
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.Date;
import net.netca.pki.encoding.json.JSON;
import net.netca.pki.encoding.json.JSONArray;
import net.netca.pki.encoding.json.JSONBoolean;
import net.netca.pki.encoding.json.JSONNumber;
import net.netca.pki.encoding.json.JSONObject;
import net.netca.pki.encoding.json.JSONString;
import net.netca.pki.encoding.asn1.pki.X509Certificate;
import net.netca.pki.Certificate;
import net.netca.pki.KeyPair;
import net.netca.pki.PkiException;
import net.netca.pki.SEStamp;
import net.netca.pki.Util;
import net.netca.pki.encoding.Base64;
import net.netca.pki.encoding.Hex;
import net.netca.pki.encoding.asn1.pki.seseal.StampBuilder;
import net.netca.pki.encoding.asn1.pki.AlgorithmIdentifier;
import net.netca.pki.encoding.asn1.pki.JCEHasher;
import net.netca.pki.encoding.asn1.pki.NetcaSigner;
import net.netca.pki.encoding.asn1.pki.seseal.ExtData;
import net.netca.pki.encoding.asn1.pki.seseal.Stamp;

public class SignSEStamp {

	public static void main(String[] args) throws Exception{
		if(args.length!=4) {
            System.out.println("Usage: java SignSEStamp issuerCertFileName pwd templateFileName stampFileName");
            return;
        }
		String issuerCertFileName=args[0];
		String pwd=args[1];
		String templateFileName=args[2];
		String stampFileName=args[3];
		JSON templateJson=getJsonFromFile(templateFileName);
		if (templateJson instanceof JSONObject==false) {
			System.out.println("模板文件不是JSON对象");
            return;
		}
		JSONObject jsonObject=(JSONObject)templateJson;
		int format=getFormat(jsonObject);
		int version=4;
		if(format!=SEStamp.GBT_38540){
			version=getVersion(jsonObject);
		}
		String vid=getVid(jsonObject);
		String id=getId(jsonObject);
		int type=getType(jsonObject);
		String name=getName(jsonObject);
		
		int certListType=-1;
		if(format==SEStamp.GMT_0031){
			certListType=SEStamp.CERTLIST_TYPE_CERT;
		} else {
			if(getUseCertDigest(jsonObject)){
				certListType=SEStamp.CERTLIST_TYPE_CERTDIGEST;
			} else {
				certListType=SEStamp.CERTLIST_TYPE_CERT;
			}
		}
		AlgorithmIdentifier hashAlgo=null;
		if(certListType==SEStamp.CERTLIST_TYPE_CERTDIGEST){
			hashAlgo=getCertHashAlgorithm(jsonObject);
		}
		
		
		ArrayList<X509Certificate> certs=getCerts(jsonObject);
		Date now=new Date();
		Date createDate=getCreateDate(jsonObject);
		if(createDate==null) {
			createDate=now;
		}
		Date validStart=getValidStart(jsonObject);
		if(validStart==null) {
			validStart=now;
		}
		Date validEnd=getValidEnd(jsonObject);
		String pictureType=getPictureType(jsonObject);
		byte[] pictureData=getPictureData(jsonObject);
		int pictureWidth=getPictureWidth(jsonObject);
		int pictureHeigth=getPictureHeigth(jsonObject);
		ArrayList<ExtData> exts=getExtDatas(jsonObject);
		String signAlgo=getSignatureAlgorithm(jsonObject);
		
		StampBuilder builder=StampBuilder.getInstance(format)
			.setVid(vid)
			.setId(id)
			.setType(type)
			.setName(name)
			.setCreateDate(createDate)
			.setValidStart(validStart)
			.setValidEnd(validEnd)
			.setPicture(pictureType, pictureData, pictureWidth, pictureHeigth);
		if(format!=SEStamp.GBT_38540){
			builder.setVersion(version);
		}
		if(certListType==SEStamp.CERTLIST_TYPE_CERT){
			for(int i=0;i<certs.size();i++) {
				builder.addCert(certs.get(i));
			}
		} else {
			for(int i=0;i<certs.size();i++) {
				builder.addCertDigest(hashAlgo, new JCEHasher(), certs.get(i));
			}
		}
		if(exts!=null) {
			for(int i=0;i<exts.size();i++) {
				builder.addExtension(exts.get(i));
			}
		}
		
		Certificate signCert=getCert(issuerCertFileName);
		KeyPair keypair=null;
		NetcaSigner signer=null;
		try {
			keypair=signCert.getKeyPair(Certificate.SEARCH_KEYPAIR_FLAG_DEVICE, Certificate.PURPOSE_SIGN,pwd);
			if(keypair==null) {
				System.out.println("获取制章人证书的私钥失败");
	            return;
			}
			signer=new NetcaSigner(keypair);
			Stamp stamp=builder.sign(new X509Certificate(signCert.derEncode()),
					signer, signAlgo);
			byte[] stampEncode=stamp.getASN1Object().encode();
			writeFile(stampFileName,stampEncode);
	        System.out.println("Success");
		} finally {
			signCert.free();
			if(keypair!=null) {
				keypair.free();
			}
			if(signer!=null) {
				signer.free();
			}
		}
	}
	
	private static void writeFile(String fileName,byte[]data) throws Exception {
		FileOutputStream out=new FileOutputStream(fileName);
		out.write(data);
		out.close();
	}
	
	private static Certificate getCert(String certFileName) throws Exception{
		byte[] certData=readFile(certFileName);
		return new Certificate(certData);
	}
	
	private static boolean getUseCertDigest(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("useCertDigest");
		if(json==null){
			return false;
		}
		if(json instanceof JSONBoolean==false) {
			throw new Exception("bad useCertDigest");
		}
		return ((JSONBoolean)json).isTrue();
		 
	}
	
	private static AlgorithmIdentifier getHashAlgorithmOid(String name) throws PkiException {
		name=name.toUpperCase();
		if(name.equals("SM3") || name.equals(AlgorithmIdentifier.SM3_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifierNullParam(AlgorithmIdentifier.SM3_OID);
		} else if(name.equals("SHA1") || name.equals("SHA-1") || name.equals(AlgorithmIdentifier.SHA1_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA1_OID);
		} else if(name.equals("SHA224") || name.equals("SHA-224") || name.equals(AlgorithmIdentifier.SHA224_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA224_OID);
		} else if(name.equals("SHA256") || name.equals("SHA-256") || name.equals(AlgorithmIdentifier.SHA256_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA256_OID);
		} else if(name.equals("SHA384") || name.equals("SHA-384") || name.equals(AlgorithmIdentifier.SHA384_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA384_OID);
		} else if(name.equals("SHA512") || name.equals("SHA-512") || name.equals(AlgorithmIdentifier.SHA512_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA512_OID);
		} else if(name.equals("MD5") || name.equals(AlgorithmIdentifier.MD5_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifierNullParam(AlgorithmIdentifier.MD5_OID);
		} else if(name.equals("SHA512/224") || name.equals("SHA-512/224") || name.equals(AlgorithmIdentifier.SHA512_224_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA512_224_OID);
		} else if(name.equals("SHA512/256") || name.equals("SHA-512/256") || name.equals(AlgorithmIdentifier.SHA512_256_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA512_256_OID);
		} else if(name.equals("SHA3-224") || name.equals(AlgorithmIdentifier.SHA3_224_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA3_224_OID);
		} else if(name.equals("SHA3-256") || name.equals(AlgorithmIdentifier.SHA3_256_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA3_256_OID);
		} else if(name.equals("SHA3-384") || name.equals(AlgorithmIdentifier.SHA3_384_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA3_384_OID);
		} else if(name.equals("SHA3-512") || name.equals(AlgorithmIdentifier.SHA3_512_OID)){
			return AlgorithmIdentifier.CreateAlgorithmIdentifier(AlgorithmIdentifier.SHA3_512_OID);
		} else {
			return AlgorithmIdentifier.CreateAlgorithmIdentifierNullParam(name);
		}
	}
	
	private static AlgorithmIdentifier getCertHashAlgorithm(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("certDigestAlgo");
		if(json instanceof JSONString==false) {
			throw new Exception("bad certDigestAlgo");
		}
		JSONString jsonString=(JSONString)json;
		return getHashAlgorithmOid(jsonString.getString()); 
	}
	
	private static ExtData getExtData(JSONObject jsonObject)throws Exception{
		JSON item=jsonObject.getValue("extnID");
		if(item instanceof JSONString==false) {
			throw new Exception("扩展OID项不是JSON字符串");
		}
		String extOid=((JSONString)item).getString();
				
		boolean critical;
		item=jsonObject.getValue("critical");
		if(item==null) {
			critical=false;
		} else if(item instanceof JSONBoolean==false) {
			throw new Exception("扩展关键性项不是JSON布尔值");
		} else {
			critical=((JSONBoolean)item).isTrue();
		}
		item=jsonObject.getValue("extnValue");
		if(item instanceof JSONString==false) {
			throw new Exception("扩展值不是JSON字符串");
		}
		byte[] value=Hex.decode(((JSONString)item).getString());
		return new ExtData(extOid,critical,value);
	}
	private static ArrayList<ExtData> getExtDatas(JSONObject jsonObject)throws Exception{
		ArrayList<ExtData> exts=new ArrayList<ExtData>();
		JSON json=jsonObject.getValue("exts");
		if(json==null) {
			return null;
		}
		if(json instanceof JSONArray==false) {
			throw new Exception("bad exts");
		}
		JSONArray jsonArray=(JSONArray)json;
		JSONObject jsonExt;
		
		ExtData ext;
		int count=jsonArray.size();
		if(count==0) {
			throw new Exception("no ext");
		}
		for(int i=0;i<count;i++) {
			json=jsonArray.get(i);
			if(json instanceof JSONObject==false) {
				throw new Exception("bad exts");
			}
			jsonExt=(JSONObject)json;
			ext=getExtData(jsonExt);
			exts.add(ext);
		}
		return exts;
	}
	
	private static String getSignatureAlgorithmOid(String name) {
		String upperName=name.toUpperCase();
		if(upperName.equals("SHA1WITHRSA")) {
			return AlgorithmIdentifier.SHA1WithRSA_OID;
		} else if(upperName.equals("SHA224WITHRSA")) {
			return AlgorithmIdentifier.SHA224WithRSA_OID;
		} else if(upperName.equals("SHA256WITHRSA")) {
			return AlgorithmIdentifier.SHA256WithRSA_OID;
		} else if(upperName.equals("SHA384WITHRSA")) {
			return AlgorithmIdentifier.SHA384WithRSA_OID;
		} else if(upperName.equals("SHA512WITHRSA")) {
			return AlgorithmIdentifier.SHA512WithRSA_OID;
		} else if(upperName.equals("SHA512/224WITHRSA")) {
			return AlgorithmIdentifier.SHA512_224WithRSA_OID;
		} else if(upperName.equals("SHA512/256WITHRSA")) {
			return AlgorithmIdentifier.SHA512_256WithRSA_OID;
		} else if(upperName.equals("SHA3-224WITHRSA")) {
			return AlgorithmIdentifier.SHA3_224WithRSA_OID;
		} else if(upperName.equals("SHA3-256WITHRSA")) {
			return AlgorithmIdentifier.SHA3_256WithRSA_OID;
		} else if(upperName.equals("SHA3-384WITHRSA")) {
			return AlgorithmIdentifier.SHA3_384WithRSA_OID;
		} else if(upperName.equals("SHA3-512WITHRSA")) {
			return AlgorithmIdentifier.SHA3_512WithRSA_OID;
		} else if(upperName.equals("SM3WITHRSA")) {
			return AlgorithmIdentifier.SM3WithRSA_OID;
		} else if(upperName.equals("SM3WITHSM2")) {
			return AlgorithmIdentifier.SM3WithSM2_OID;
		} else if(upperName.equals("DSAWITHSHA1")) {
			return AlgorithmIdentifier.DSAWithSHA1_OID;
		} else if(upperName.equals("DSAWITHSHA224")) {
			return AlgorithmIdentifier.DSAWithSHA224_OID;
		} else if(upperName.equals("DSAWITHSHA256")) {
			return AlgorithmIdentifier.DSAWithSHA256_OID;
		} else if(upperName.equals("ECDSAWITHSHA1")) {
			return AlgorithmIdentifier.ECDSAWithSHA1_OID;
		} else if(upperName.equals("ECDSAWITHSHA224")) {
			return AlgorithmIdentifier.ECDSAWithSHA224_OID;
		} else if(upperName.equals("ECDSAWITHSHA256")) {
			return AlgorithmIdentifier.ECDSAWithSHA256_OID;
		} else if(upperName.equals("ECDSAWITHSHA384")) {
			return AlgorithmIdentifier.ECDSAWithSHA384_OID;
		} else if(upperName.equals("ECDSAWITHSHA512")) {
			return AlgorithmIdentifier.ECDSAWithSHA512_OID;
		} else if(upperName.equals("ECDSAWITHSHA3-224")) {
			return AlgorithmIdentifier.ECDSAWithSHA3_224_OID;
		} else if(upperName.equals("ECDSAWITHSHA3-256")) {
			return AlgorithmIdentifier.ECDSAWithSHA3_256_OID;
		} else if(upperName.equals("ECDSAWITHSHA3-384")) {
			return AlgorithmIdentifier.ECDSAWithSHA3_384_OID;
		} else if(upperName.equals("ECDSAWITHSHA3-512")) {
			return AlgorithmIdentifier.ECDSAWithSHA3_512_OID;
		} 
		return name;
	}
	private static String getSignatureAlgorithm(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("signAlgo");
		if(json instanceof JSONString==false) {
			throw new Exception("bad signAlgo");
		}
		JSONString jsonString=(JSONString)json;
		return getSignatureAlgorithmOid(jsonString.getString()); 
	}
	private static int getPictureHeigth(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("picture");
		if(json instanceof JSONObject==false) {
			throw new Exception("bad picture");
		}
		JSONObject picture=(JSONObject)json;
		json=picture.getValue("heigth");
		if(json instanceof JSONNumber==false) {
			throw new Exception("bad picture heigth");
		}
		JSONNumber jsonNumber=(JSONNumber)json;
		return jsonNumber.getNumber().intValue();
	}
	
	private static int getPictureWidth(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("picture");
		if(json instanceof JSONObject==false) {
			throw new Exception("bad picture");
		}
		JSONObject picture=(JSONObject)json;
		json=picture.getValue("width");
		if(json instanceof JSONNumber==false) {
			throw new Exception("bad picture width");
		}
		JSONNumber jsonNumber=(JSONNumber)json;
		return jsonNumber.getNumber().intValue();
	}
	
	private static byte[] getPictureData(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("picture");
		if(json instanceof JSONObject==false) {
			throw new Exception("bad picture");
		}
		JSONObject picture=(JSONObject)json;
		json=picture.getValue("data");
		if(json instanceof JSONString==false) {
			throw new Exception("bad picture data");
		}
		JSONString jsonString=(JSONString)json;
		return Base64.decode(true,jsonString.getString());
	}
	
	private static String getPictureType(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("picture");
		if(json instanceof JSONObject==false) {
			throw new Exception("bad picture");
		}
		JSONObject picture=(JSONObject)json;
		json=picture.getValue("type");
		if(json instanceof JSONString==false) {
			throw new Exception("bad picture type");
		}
		JSONString jsonString=(JSONString)json;
		return jsonString.getString();
	}
	
	private static Date getValidEnd(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("validEnd");
		if(json instanceof JSONString==false) {
			throw new Exception("bad validEnd");
		}
		JSONString jsonString;
		jsonString=(JSONString)json;
		return Util.UTCTimeDecode(jsonString.getString().getBytes("UTF-8"));
	}
	
	private static Date getValidStart(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("validStart");
		if(json==null) {
			return null;
		}
		if(json instanceof JSONString==false) {
			throw new Exception("bad validStart");
		}
		JSONString jsonString;
		jsonString=(JSONString)json;
		return Util.UTCTimeDecode(jsonString.getString().getBytes("UTF-8"));
	}
	
	private static Date getCreateDate(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("createDate");
		if(json==null) {
			return null;
		}
		if(json instanceof JSONString==false) {
			throw new Exception("bad createDate");
		}
		JSONString jsonString;
		jsonString=(JSONString)json;
		return Util.UTCTimeDecode(jsonString.getString().getBytes("UTF-8"));
	}
	
	private static ArrayList<X509Certificate> getCerts(JSONObject jsonObject)throws Exception{
		ArrayList<X509Certificate> certs=new ArrayList<X509Certificate>();
		JSON json=jsonObject.getValue("certs");
		if(json instanceof JSONArray==false) {
			throw new Exception("bad certs");
		}
		JSONArray jsonArray=(JSONArray)json;
		JSONString jsonString;
		X509Certificate cert;
		int count=jsonArray.size();
		if(count==0) {
			throw new Exception("no cert");
		}
		for(int i=0;i<count;i++) {
			json=jsonArray.get(i);
			if(json instanceof JSONString==false) {
				throw new Exception("bad certs");
			}
			jsonString=(JSONString)json;
			cert=new X509Certificate(jsonString.getString());
			certs.add(cert);
		}
		return certs;
	}
	private static String getName(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("name");
		if(json instanceof JSONString==false) {
			throw new Exception("bad name");
		}
		return ((JSONString)json).getString();
	}
	
	private static int getType(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("type");
		if(json instanceof JSONNumber==false) {
			throw new Exception("bad type");
		}
		return ((JSONNumber)json).getNumber().intValue();
	}
	
	private static String getVid(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("vid");
		if(json instanceof JSONString==false) {
			throw new Exception("bad vid");
		}
		return ((JSONString)json).getString();
	}
	
	private static int getFormat(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("format");
		if(json==null){
			return SEStamp.GBT_38540;
		}
		if(json instanceof JSONString==false) {
			throw new Exception("bad format");
		}
		String str=((JSONString)json).getString();
		if(str.equals("GB/T 38540")){
			return SEStamp.GBT_38540;
		} else if(str.equals("GM/T 0031")){
			return SEStamp.GMT_0031;
		} else {
			throw new Exception("bad format");
		}
	}
	
	private static int getVersion(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("version");
		if(json instanceof JSONNumber==false) {
			throw new Exception("bad version");
		}
		return ((JSONNumber)json).getNumber().intValue();
	}
	
	private static String getId(JSONObject jsonObject)throws Exception{
		JSON json=jsonObject.getValue("id");
		if(json instanceof JSONString==false) {
			throw new Exception("bad id");
		}
		return ((JSONString)json).getString();
	}
	
	private static JSON getJsonFromFile(String fileName) throws Exception{
		byte[] data=readFile(fileName);
		String str;
		if(data.length>=3 && (data[0]&0xFF)==0xEF && (data[1]&0xFF)==0xBB && (data[2]&0xFF)==0xBF) {
			str=new String(data,3,data.length-3,"UTF-8");
		} else {
			str=new String(data,"UTF-8");
		}
		JSON json=JSON.decode(str);
		if(json==null) {
			throw new Exception("bad json file:" + fileName);
		}
		return json;
	}
	
	private static byte[] readFile(String fileName) throws Exception{
		File file=new File(fileName);
		
		long fileLen=file.length();
		if(fileLen>=java.lang.Integer.MAX_VALUE) {
			throw new Exception("file too big:"+ fileName);
		}
		int len=(int)fileLen;
		byte[] data=new byte[len];
		FileInputStream fIn=new FileInputStream(file);
		
		try{
			if (fIn.read(data)!=len) {
				throw new Exception("read file fail:" + fileName);
			}
			return data;
		} finally {
			fIn.close();
		}
	}
}

File: 例子/lua的例子/SignSEStamp.lua
local CryptCore=require("CryptCore")

function ToANSIString(utf8String)
    return CryptCore.UTF8ToACP(utf8String)
end


function Usage(filename)
    print(string.format("Usage: lua %s issuerCertFileName templateFileName stampFileName [pwd]",filename))
    print(ToANSIString("参数"))
    print(ToANSIString("\tissuerCertFileName：制章人证书文件名"))
    print(ToANSIString("\ttemplateFileName：模板文件名"))
    print(ToANSIString("\tstampFileName：输出的电子印章的文件名"))
    print(ToANSIString("\tpwd：使用私钥需要的密码"))
    print(ToANSIString("说明"))
    print(ToANSIString("\t根据templateFileName文件的信息进行签名产生电子印章，结果保存到stampFileName文件中"))
end

function GetFileConetent(fileName)
    local fIn=io.open(fileName,"rb")
    if fIn==nil then
        return nil
    end
    local data=fIn:read("*a")
    fIn:close()
    
    return data
end

function Save(fileName,data)
    local fOut=io.open(fileName,"wb")
    if fOut==nil then
        return false
    end
    fOut:write(data)
    fOut:close()    
    return true
end


function GetSignAlgo(signAlgo)
    if signAlgo==nil then
        return nil
    end
    signAlgo=signAlgo:upper()
    if signAlgo== "SHA1WITHRSA" or signAlgo=="SHA-1WITHRSA" or signAlgo== "SHA1WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.5"then
        return CryptCore.Const.SHA1WITHRSA
    elseif  signAlgo== "SHA224WITHRSA" or signAlgo== "SHA224WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.14" then
        return CryptCore.Const.SHA224WITHRSA
    elseif  signAlgo== "SHA256WITHRSA" or signAlgo== "SHA256WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.11" then
        return CryptCore.Const.SHA256WITHRSA
    elseif  signAlgo== "SHA384WITHRSA" or signAlgo== "SHA384WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.12" then
        return CryptCore.Const.SHA384WITHRSA
    elseif  signAlgo== "SHA512WITHRSA" or signAlgo== "SHA512WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.13" then
        return CryptCore.Const.SHA512WITHRSA
    elseif  signAlgo== "MD5WITHRSA" or signAlgo== "MD5WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.4" then
        return CryptCore.Const.MD5WITHRSA
    elseif  signAlgo== "SHA512/224WITHRSA" or signAlgo== "SHA512/224WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.15" then
        return 34
    elseif  signAlgo== "SHA512/256WITHRSA" or signAlgo== "SHA512/256WITHRSAENCRYPTION" or signAlgo=="1.2.840.113549.1.1.16" then
        return 35
    elseif  signAlgo== "SHA3-224WITHRSA" or signAlgo== "SHA3-224WITHRSAENCRYPTION" or signAlgo=="2.16.840.1.101.3.4.3.13" then
        return 36
    elseif  signAlgo== "SHA3-256WITHRSA" or signAlgo== "SHA3-256WITHRSAENCRYPTION" or signAlgo=="2.16.840.1.101.3.4.3.14" then
        return 37
    elseif  signAlgo== "SHA3-384WITHRSA" or signAlgo== "SHA3-384WITHRSAENCRYPTION" or signAlgo=="2.16.840.1.101.3.4.3.15" then
        return 38
    elseif  signAlgo== "SHA3-512WITHRSA" or signAlgo== "SHA3-512WITHRSAENCRYPTION" or signAlgo=="2.16.840.1.101.3.4.3.16" then
        return 39
    elseif  signAlgo== "SM3WITHSM2" or signAlgo=="1.2.156.10197.1.501" then
        return CryptCore.Const.SM3WITHSM2
    elseif  signAlgo== "ECDSAWITHSHA1" or signAlgo=="1.2.840.10045.4.1" then
        return CryptCore.Const.ECDSAWITHSHA1
    elseif  signAlgo== "ECDSAWITHSHA224" or signAlgo=="1.2.840.10045.4.3.1" then
        return CryptCore.Const.ECDSAWITHSHA224
    elseif  signAlgo== "ECDSAWITHSHA256" or signAlgo=="1.2.840.10045.4.3.2"  then
        return CryptCore.Const.ECDSAWITHSHA256
    elseif  signAlgo== "ECDSAWITHSHA384" or signAlgo=="1.2.840.10045.4.3.3" then
        return CryptCore.Const.ECDSAWITHSHA384
    elseif  signAlgo== "ECDSAWITHSHA512" or signAlgo=="1.2.840.10045.4.3.4" then
        return CryptCore.Const.ECDSAWITHSHA512
    elseif  signAlgo== "ECDSAWITHSHA3-224" or signAlgo=="2.16.840.1.101.3.4.3.9" then
        return 40
    elseif  signAlgo== "ECDSAWITHSHA3-256" or signAlgo=="2.16.840.1.101.3.4.3.10" then
        return 41
    elseif  signAlgo== "ECDSAWITHSHA3-384" or signAlgo=="2.16.840.1.101.3.4.3.11" then
        return 42
    elseif  signAlgo== "ECDSAWITHSHA3-512" or signAlgo=="2.16.840.1.101.3.4.3.12" then
        return 43
    else
        return nil;
    end
end

function GetHashAlgo(algo)
    if algo==nil then
        return nil
    end
    algo=algo:upper()
    if algo== "SHA1" or algo=="SHA-1" or algo=="1.3.14.3.2.26" then
        return CryptCore.Const.SHA1
    elseif  algo== "SHA224" or algo=="SHA-224" or algo=="2.16.840.1.101.3.4.2.4" then
        return CryptCore.Const.SHA224
    elseif  algo== "SHA256" or algo=="SHA-256" or algo=="2.16.840.1.101.3.4.2.1" then
        return CryptCore.Const.SHA256
    elseif  algo== "SHA384" or algo=="SHA-384" or algo=="2.16.840.1.101.3.4.2.2" then
        return CryptCore.Const.SHA384
    elseif  algo== "SHA512" or algo=="SHA-512" or algo=="2.16.840.1.101.3.4.2.3" then
        return CryptCore.Const.SHA512
    elseif  algo== "MD5" or algo=="1.2.840.113549.2.5" then
        return CryptCore.Const.MD5
    elseif  algo== "SHA512/224" or algo== "SHA-512/224" or algo=="2.16.840.1.101.3.4.2.5" then
        return CryptCore.Const["SHA-512/224"]
    elseif  algo== "SHA512/256" or algo== "SHA-512/256" or algo=="2.16.840.1.101.3.4.2.6" then
        return CryptCore.Const["SHA-512/256"]
    elseif  algo== "SHA3-224" or algo=="2.16.840.1.101.3.4.2.7" then
        return CryptCore.Const["SHA3-224"]
    elseif  algo== "SHA3-256" or algo=="2.16.840.1.101.3.4.2.8" then
        return CryptCore.Const["SHA3-256"]
    elseif  algo== "SHA3-384" or algo=="2.16.840.1.101.3.4.2.9" then
        return CryptCore.Const["SHA3-384"]
    elseif  algo== "SHA3-512" or algo=="2.16.840.1.101.3.4.2.10" then
        return CryptCore.Const["SHA3-512"]
    elseif  algo== "SM3" or algo=="1.2.156.10197.1.401" then
        return CryptCore.Const.SM3
    else
        return nil;
    end
end


if #arg~=4 and #arg~=3 then
    Usage(arg[0])
    os.exit(-1)
end

local issuerCertFileName=arg[1]
local templateFileName=arg[2]
local stampFileName=arg[3]
local cert=GetFileConetent(issuerCertFileName)
if cert==nil then
    print(ToANSIString("读证书文件失败"));
    os.exit(-1)
end

local certObj=CryptCore.NewCert(cert)
if certObj==nil then
    print(ToANSIString("不是证书文件"));
    os.exit(-1)
end


if certObj:Match("CertType='Signature'")==false then
    certObj:Free()
    print(ToANSIString("不是签名证书"));
    os.exit(-1)
end

local keypair
local pwd=nil
--如果需要从证书库里获取私钥，则使用CryptCore.Const.SEARCH_KEYPAIR_FLAG_CURRENTUSER代替CryptCore.Const.SEARCH_KEYPAIR_FLAG_DEVICE
if #arg==4 then
    pwd=arg[4]
    keypair=certObj:GetPrivateKey(CryptCore.Const.SEARCH_KEYPAIR_FLAG_DEVICE,CryptCore.Const.CERT_PURPOSE_SIGN,pwd)
else
    keypair=certObj:GetPrivateKey(CryptCore.Const.SEARCH_KEYPAIR_FLAG_DEVICE,CryptCore.Const.CERT_PURPOSE_SIGN)
end
if keypair==nil then
	certObj:Free()
    if pwd==nil then
        print(ToANSIString("证书没有私钥"));
    else
        print(ToANSIString("证书没有私钥，或者密码错误"));
    end
    os.exit(-1)
end
keypair:Free()


local jsonData=GetFileConetent(templateFileName)
if jsonData==nil then
	certObj:Free()
    print(ToANSIString("读模板文件失败"));
    os.exit(-1)
end

local json=CryptCore.Json.Decode(jsonData)
if json==nil then
	certObj:Free()
	print(ToANSIString("模板文件不是JSON"));
    os.exit(-1)
end

local format
local formatStr=json.format
if formatStr==nil or formatStr=="GB/T 38540" then
    format=CryptCore.Const["GB/T 38540"]
elseif formatStr=="GM/T 0031" then
    format=CryptCore.Const["GM/T 0031"]
else
    certObj:Free()
	print(ToANSIString("模板文件的format出错"));
    os.exit(-1)
end

local stamp=assert(CryptCore.NewStamp(format))
if format==CryptCore.Const["GM/T 0031"] then
    stamp:SetVersion(json.version)
end
stamp:SetVid(json.vid)
stamp:SetId(json.id)
stamp:SetType(json.type)
stamp:SetName(json.name)

local useCertDigest=false
local hashAlgo
if format==CryptCore.Const["GM/T 0031"] then
    useCertDigest=false
else 
    useCertDigest=json.useCertDigest
    if useCertDigest then
        hashAlgo=GetHashAlgo(json.certDigestAlgo)
        if hashAlgo==nil then
            stamp:Free()
            certObj:Free()
            print(ToANSIString("不支持的Hash算法"));
            os.exit(-1)
        end
    end
end

certs=json.certs
local certObj2
for i,cert in ipairs(certs) do
    certObj2=CryptCore.NewCert(cert)
    if certObj2==nil then
        stamp:Free()
        certObj:Free()
        print(ToANSIString("签章证书格式错误"));
        os.exit(-1)
    end
    if useCertDigest then
        stamp:AddCertDigest(hashAlgo,certObj2)
    else
        stamp:AddCert(certObj2)
    end
    certObj2:Free()
end

stamp:SetCreateDate(json.createDate)
stamp:SetValidStart(json.validStart)
stamp:SetValidEnd(json.validEnd)

local picture=json.picture
stamp:SetPicture(picture.type,picture.data,picture.width,picture.heigth)

local exts=json.exts
if exts~=nil then
    for i,ext in ipairs(exts) do
        stamp:AddExtension(ext.extnID,ext.critical,CryptCore.HexDecode(ext.extnValue))
    end
end

local signAlgo=GetSignAlgo(json.signAlgo)
if signAlgo==nil then
    print(ToANSIString("不支持的签名算法"));
    os.exit(-1)
end
local stampEncode=stamp:Sign(certObj,signAlgo)
stamp:Free()
certObj:Free()

if stampEncode==nil then
    print(ToANSIString("签名失败"));
    os.exit(-1)
end

if not Save(stampFileName,stampEncode) then
    print(ToANSIString("保存电子印章失败"));
    os.exit(-1)
end
print(ToANSIString("签电子印章成功"));